File: src/main/java/com/nurflugel/dependencyvisualizer/io/PersonAdapter.java
package com.nurflugel.dependencyvisualizer.io;

import com.google.gson.*;
import com.nurflugel.dependencyvisualizer.data.pojos.Person;
import java.lang.reflect.Type;

/** Created by douglas_bullard on 1/18/16. */
@SuppressWarnings("Duplicates")
public class PersonAdapter                // implements JsonSerializer<Person>, JsonDeserializer<Person>
{
  // @Override
  public JsonElement serialize(Person src, Type typeOfSrc, JsonSerializationContext context)
  {
    JsonObject result = new JsonObject();

    result.add("type", new JsonPrimitive(src.getClass().getSimpleName()));
    result.add("properties", context.serialize(src, src.getClass()));

    return result;
  }

  // @Override
  public Person deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException
  {
    JsonObject  jsonObject = json.getAsJsonObject();
    JsonElement element    = jsonObject.get("type");

    if (element != null)
    {
      element = jsonObject.get("properties");
    }

    String className = element.getAsString();

    try
    {
      String theClassName = "com.nurflugel.dependencyvisualizer.data.pojos." + className;

      System.out.println("theClassName = " + theClassName);

      Class<?> classInstance = Class.forName(theClassName);

      return context.deserialize(element, classInstance);
    }
    catch (ClassNotFoundException cnfe)
    {
      throw new JsonParseException("Unknown element type: " + className, cnfe);
    }
  }
}

File: src/main/kotlin/com/nurflugel/dependencyvisualizer/io/PersonAdapter.kt
package com.nurflugel.dependencyvisualizer.io

import com.google.gson.*
import com.nurflugel.dependencyvisualizer.data.pojos.Person
import java.lang.reflect.Type

/** Created by douglas_bullard on 1/18/16.
 *
 * todo delete me
 * */
class PersonAdapter // implements JsonSerializer<Person>, JsonDeserializer<Person>

{
    // @Override
    fun serialize(src: Person, typeOfSrc: Type?, context: JsonSerializationContext): JsonElement {
        val result = JsonObject()

        result.add("type", JsonPrimitive(src.javaClass.simpleName))
        result.add("properties", context.serialize(src, src.javaClass))

        return result
    }

    // @Override
    @Throws(JsonParseException::class)
    fun deserialize(json: JsonElement, typeOfT: Type?, context: JsonDeserializationContext): Person {
        val jsonObject = json.asJsonObject
        var element = jsonObject["type"]

        if (element != null) {
            element = jsonObject["properties"]
        }

        val className = element!!.asString

        try {
            val theClassName = "com.nurflugel.dependencyvisualizer.data.pojos.$className"

            println("theClassName = $theClassName")

            val classInstance = Class.forName(theClassName)

            return context.deserialize(element, classInstance)
        } catch (cnfe: ClassNotFoundException) {
            throw JsonParseException("Unknown element type: $className", cnfe)
        }
    }
}
